<% brew::brew(get.path2template("page.header")) %>

<%
# Variables:
#  table.html - table with the data per sample + gene information
#  groups.l - samples per group
#  html.cols2sel - column names of gene name + samples
#  coloursbygroup
#  quant.data 
#  quant.data.norm 
#  quant.data.norm.group 
#  quant.data.group
%>

<%# Comment -- wrap up everything in a div %>
<DIV name='de_res'>
<H1><%=title%></H1>

<H2 class="section">Quantification</H2>

<UL>
  <LI><%=pprint.feat(opt$feature,cap=T)%> level analysis
  <LI>Metric: <%=pprint.metric(opt$metric)%>
</UL>
<DIV class="de_report_bar">Set of <%=pprint.feat(opt$feature,cap=F)%>s considered: <b><%=filt.name%></b>
<DIV class="source_filter"> <%=sources.menu%>
</DIV>
</DIV>

<OL class="contents"> Contents
  <LI> <A href="#data">TSV file with the data</A>
  <LI> <A href="#gl">Group level analysis</A>
  <LI> <A href="#sl">Library level analysis</A>
</OL>

<HR/>

<%
tsv.download.files <- c(paste(basename(tsv.file.ref),".gz",sep=""))
names(tsv.download.files) <- "TSV"
tsv.download.html <- html.download.bar(tsv.download.files)
file.prefix <- paste(file.prefix,opt$metric,sep="_")
# global variables
# clustering
library(agricolae)
d <- "euclidean"
m <- "complete"
nboot<-100
# when should the values be log2
dolog2 <- function(d) {
  if(opt$metric=="raw" || opt$metric=="nlib" ) {
    return(log2(d+1))
  }
  return(d)
}
# enough data to generate a set of trees
do_clustering<-function(m) {
  if ( length(colnames(m)) >= 3 && nrow(m)>=3 )
    return(TRUE)
  return(FALSE)
}
# enough data to generate a heatmap
do_heatmap<-function(m) {
  if ( length(colnames(m)) >= 2 && nrow(m)>=5 )
    return(TRUE)
  return(FALSE)
}
do_cor_heatmap<-function(m) {
  if ( length(colnames(m)) >= 2 && nrow(m)>=30 )
    return(TRUE)
  return(FALSE)
}

%>

 <H2 class="section">Data<A NAME="data"></H2>
 <%=tsv.download.html%>
 <p> Number of <%=pprint.feat(opt$feature)%>s:<%=nrow(quant.data)%>

  
<H2 class="section">Group level analysis<A NAME="gl"></H2>


<H3><%=irap.ctr("Figure")%>: Correlation scatterplot</H3>

<%
if (ncol(quant.data)>=2 && ncol(quant.data)<12 && nrow(quant.data)>20 ) {
  pdebug("Correlation scatterplots\n")    
  p <- gen.plot2report(filename=paste(file.prefix,"_group_pairs.png",sep=""),
                       width=600,
                       height=600,
                       caption=pprint.metric(opt$metric),
                       to.plot=function() {
                         pairs(dolog2(quant.data.group$mean),  lower.panel=panel.smooth, upper.panel=my.panel.cor,pch=".", main=paste("",sep=""))
                       }
                       ) 
  html <- p$html

  irap.output.html(html,paste("Pairwise comparison across all groups. Panels above the diagonal show the Pearson correlation between two groups. Panels below the diagonal show the log2 expression of each ",pprint.feat(opt$feature,cap=F)," for each pair of groups. A regression  (computed using the R lowess function) line is plotted in red.",sep=""))

} else {
  cat("Insufficient data or plot not suitable for the available data")
}
%>

<H3><%=irap.ctr("Figure")%>: Clustering </H3>
<%
  ####################
  # clustering
  if ( do_clustering(quant.data.group$mean)){
    pdebug("Clustering\n")    
    p <- gen.plot2report(filename=paste(file.prefix,"_group_clus.png",sep=""),
                    width=500,
                    height=500,
                         caption=paste("Clustering (distance=",d,"; method=",m,";restarts= ",nboot,"): ",pprint.metric(opt$metric),sep=""),
                    to.plot=function() {
                      con<-irap.consensus(as.data.frame(t(dolog2(quant.data.group$mean))),distance=d,method=m, nboot=nboot,duplicate=TRUE,xlab="",yaxt='n', ann=FALSE,leafsColours=coloursbygroup)      
                    }
                    )
    html <- p$html
    
    if ( opt$metric=="raw") {  
       p <- gen.plot2report(filename=paste(file.prefix,"_group_clus_norm.png",sep=""),
                    width=500,
                    height=500,
                    caption=paste("Clustering (distance=",d,"; method=",m,";restarts= ",nboot,"): ",pprint.metric("nlib"),sep=""),
                    to.plot=function() {
                      con<-irap.consensus(as.data.frame(t(dolog2(quant.data.norm.group$mean))),distance=d,method=m, nboot=nboot,duplicate=TRUE, xlab="",yaxt='n', ann=FALSE,leafsColours=coloursbygroup)      
                    }
                    )
       html <- paste(html,p$html,sep=" ")
     }
    irap.output.html(html,paste("Clustering of the groups by taking the mean values of the libraries belonging to each group. The plot shows the consensus of ",nboot," trees computed using the ",m," method and distance ",d,".",sep=""))
     
  } else {
    cat("<p>Insufficient data to generate the clustering</p>")
    pdebug("Insufficient data to generate the clustering")
  }
%>

<H3><%=irap.ctr("Figure")%>: Heatmap</H3>
<%
# all genes/exons/trans
pdebug("Heatmaps(group)")
nlim <- 100
if (do_heatmap(quant.data.group$mean)) {
  p <- gen.plot2report(filename=paste(file.prefix,"allg","_group_heatmap_s.png",sep=""),
                       width=600,
                       height=600,
                       caption=pprint.metric(opt$metric),
                       to.plot=function() {
                         quant.heatmap(dolog2(quant.data.group$mean),do.cor=T,nlim=NA,ColSideColors=coloursbygroup[colnames(quant.data.group$mean)],srtCol=45)
                       }
                    )
  
  html <- p$html

  if ( opt$metric=="raw") {  
    p <- gen.plot2report(filename=paste(file.prefix,"allg_group_heatmap_s.png",sep=""),
                       width=600,
                       height=600,
                       caption=pprint.metric("nlib"),
                       to.plot=function() {
                         quant.heatmap(quant.data.norm.group$mean,do.cor=T,nlim=NA,ColSideColors=coloursbygroup[colnames(quant.data.norm.group$mean)],srtCol=45)
                       }
                    )
  
    html <- paste(html,p$html,sep=" ")
  }
  heatmap.cor.group.info <- paste("Heatmap(s) with the Pearson correlation of ",pprint.feat(opt$feature)," expression values between each  group of libraries. The expression value for a ",pprint.feat(opt$feature)," in a group corresponds to the mean values observed in the libraries of the group.",sep="")
  heatmap.group.info <- paste("First heatmap is based on the Pearson correlation of ",pprint.feat(opt$feature)," expression values between each group of libraries. Second heatmap is based  on the ",pprint.feat(opt$feature),"s' expression values for each library. The expression value for a ",pprint.feat(opt$feature)," in a group corresponds to the mean values observed in the groups' libraries. Only the subset with ",nlim," genes/exons/transcripts with largest variance was considered.",sep="")
  heatmap.sample.info <- paste("Heatmaps based on ",pprint.feat(opt$feature)," expression values for each library. Only the subset with ",nlim," genes/exons/transcripts with largest variance was considered.",sep="")

  #heatmap.cor <- "Heatmap with the Pearson correlation of the expression by a group of libraries (the expression values by group corresponds to the mean values observed in the libraries of the group)."
  irap.output.html(html,paste(heatmap.cor.group.info,sep=""))

} else {
  cat("<p>Insufficient data to generate heatmap</p>")
  pdebug("Insufficient data to generate heatmap")
}
%>

<H3><%=irap.ctr("Figure")%>: Heatmap (subset of <%=pprint.feat(opt$feature,cap=F)%>s)</H3>
<%
pdebug("Heatmaps(group)")
nlim <- 100
if (do_heatmap(quant.data.group$mean)) {
  p <- gen.plot2report(filename=paste(file.prefix,nlim,"_group_heatmap_s1.png",sep=""),
                       width=600,
                       height=600,
                       caption=paste("Heatmap: ",pprint.metric(opt$metric)," for (",nlim," ) ",pprint.feat(opt$feature,cap=F)," with larger variance",sep=""),
                       to.plot=function() {
                         quant.heatmap(dolog2(quant.data.group$mean),do.cor=T,nlim=nlim,ColSideColors=coloursbygroup[colnames(quant.data.group$mean)],srtCol=45)
                       }
                    )
  html <- p$html
  p <- gen.plot2report(filename=paste(file.prefix,nlim,"_group_heatmap_s2.png",sep=""),
                       width=600,
                       height=600,
                       caption=paste("Heatmap: ",pprint.metric(opt$metric)," (",nlim," ",pprint.feat(opt$feature,cap=T),"s with larger variance)",sep=""),
                       to.plot=function() {
                         #
                         quant.heatmap(dolog2(quant.data.group$mean),do.cor=F,nlim=nlim,ColSideColors=coloursbygroup[colnames(quant.data.group$mean)],srtCol=45)
                       }
                    )
  html <- paste(html,p$html,sep=" ")
  if ( opt$metric=="raw") {  
    p <- gen.plot2report(filename=paste(file.prefix,nlim,"_group_heatmap_s3.png",sep=""),
                       width=600,
                       height=600,
                       caption=paste("Heatmap: ",pprint.metric("nlib")," (",nlim," ",pprint.feat(opt$feature,cap=T),"s with larger variance)",sep=""),
                       to.plot=function() {
                         quant.heatmap(dolog2(quant.data.norm.group$mean),do.cor=T,nlim=nlim,ColSideColors=coloursbygroup[colnames(quant.data.norm.group$mean)],srtCol=45)
                       }
                    )
    
    html <- paste(html,p$html,sep=" ")
        
    p <- gen.plot2report(filename=paste(file.prefix,nlim,"_group_heatmap_s4.png",sep=""),
                     width=600,
                     height=600,
                     caption=paste("Heatmap: ",pprint.metric("nlib")," (",nlim," ",pprint.feat(opt$feature,cap=T),"s with larger variance)",sep=""),
                     to.plot=function() {
                       quant.heatmap(dolog2(quant.data.norm.group$mean),do.cor=F,nlim=nlim,ColSideColors=coloursbygroup[colnames(quant.data.norm.group$mean)],srtCol=45)
                    }
                    )
  }
  html <- paste(html,p$html,sep=" ")
  
  irap.output.html(html,paste(heatmap.group.info,sep=""))
} else {
    cat("<p>Insufficient data to generate heatmap</p>")
    pdebug("Insufficient data to generate heatmap")
  }
%>

  
<H2 class="section">Library level analysis<A NAME="sl"/></H2>

<!-- libraries -->
<!-- Correlations -->

<H3><%=irap.ctr("Figure")%>: Correlation between the libraries</H3>
<%
pdebug("Sample correlations")
pdebug("#Vars=",ncol(quant.data))
# split the plot if two many libraries
if (ncol(quant.data)>=10 ) {
  # show by group
  for (g in names(groups.l)) {
    #g <- names(groups.l)[1]
    cols <- groups.l[[g]]
    pdebug("Scatterplot for group ",g)
    quant.data.sel <- quant.data[,cols]
    if (length(cols)>1 &&  nrow(quant.data.sel)>20 ) {
      p <- gen.plot2report(filename=paste(file.prefix,"_sample_pairs_",g,".png",sep=""),
                           width=600,
                           height=600,
                           to.plot=function() {
                             pairs(dolog2(quant.data.sel),lower.panel=panel.smooth, upper.panel=my.panel.cor,pch=".", main=paste(g,"",sep=""))
                           }
                           )
      irap.output.html(p$html,paste("Pearson correlation",sep=""))
    }
  }
} else {
  if (ncol(quant.data)>=2  && nrow(quant.data)>20 ) {
    p <- gen.plot2report(filename=paste(file.prefix,"_sample_pairs.png",sep=""),
                         width=600,
                         height=600,
                         to.plot=function() {
			   pairs(dolog2(quant.data),  lower.panel=panel.smooth, upper.panel=my.panel.cor,pch=".", main=paste("",sep=""))
                         }
                         )
    irap.output.html(p$html,paste("Pearson correlation",sep=""))
  } else {			   
    cat("<p>Insufficient data to generate plot</p>")
    pdebug("Insufficient data to generate pairs plot")   
  }
}
%>
  
<!-- clustering -->
<H3><%=irap.ctr("Figure")%>: Clustering (libraries)</H3>
<%
  ####################
  # clustering

  if ( do_clustering(quant.data))   {
    pdebug("Clustering (libraries)\n")    
    p <- gen.plot2report(filename=paste(file.prefix,"_sample_clus.png",sep=""),
                    width=500,
                    height=500,
                    caption=paste("Clustering (distance=",d,"; method=",m,";restarts= ",nboot,"):",pprint.metric(opt$metric),sep=""),
                    to.plot=function() {
                      con<-irap.consensus(as.data.frame(t(dolog2(quant.data))),distance=d,method=m, nboot=nboot,duplicate=TRUE,xlab="",yaxt='n', ann=FALSE, leafsColours=coloursbysample)
                    }
                    )
    html <- p$html
    if ( opt$metric=="raw") {  
      p <- gen.plot2report(filename=paste(file.prefix,"_sample_clus_norm.png",sep=""),
                           width=500,
                           height=500,
                           caption=paste("Clustering (distance=",d,"; method=",m,";restarts= ",nboot,"): ",pprint.metric("nlib"),sep=""),
                           to.plot=function() {
                             con<-irap.consensus(as.data.frame(t(dolog2(quant.data.norm))),distance=d,method=m, nboot=nboot,duplicate=TRUE,xlab="",yaxt='n', ann=FALSE,leafsColours=coloursbysample)      
                           }
                           )
      html <- paste(html,p$html, sep=" ")
    }
    irap.output.html(html,paste("Clustering of all libraries. The plot shows the consensus of ",nboot," trees computed using the ",m," method and distance ",d,".",sep=""))
  } else {
    cat("<p>Insufficient data to generate the clustering</p>")
    pdebug("Insufficient data to generate the clustering")
  }
%>

<!-- heatmap -->
<%
  nlim <- 100
%>
<H3><%=irap.ctr("Figure")%>: Heatmap for <%=nlim%> <%=pprint.feat(opt$feature,cap=T)%>s with larger variance (libraries)</H3>
<%
pdebug("Heatmap")
# cor for the libraries
if (do_heatmap(quant.data)) {
  p <- gen.plot2report(filename=paste(file.prefix,"_sample_heatmap_s1",nlim,".png",sep=""),
                     width=800,
                     height=800,
                     caption=paste("Heatmap: ",pprint.metric(opt$metric),"(",nlim," ",pprint.feat(opt$feature,cap=F),"s with larger variance)",sep=""),
                     to.plot=function() {
                       quant.heatmap(dolog2(quant.data),do.cor=T,nlim=nlim,ColSideColors=coloursbygroup[conds])
                    }
                    )
										       
  html <- p$html
  p <- gen.plot2report(filename=paste(file.prefix,"_sample_heatmap_g1",nlim,".png",sep=""),
                       width=800,
                       height=800,
                       caption=paste("Heatmap: ",pprint.metric(opt$metric),"(",nlim," ",pprint.feat(opt$feature,cap=F),"s with larger variance)",sep=""),
                       to.plot=function() {
                         quant.heatmap(dolog2(quant.data),do.cor=F,nlim=nlim,ColSideColors=coloursbygroup[conds])
                       }
                       )

  html <- paste(html,p$html,sep=" ")
  if ( opt$metric=="raw") {  
    p <- gen.plot2report(filename=paste(file.prefix,"_norm_sample_heatmap_s2",nlim,".png",sep=""),
                         width=800,
                         height=800,
                         caption=paste("Heatmap: ",pprint.metric("nlib")," (",nlim," ",pprint.feat(opt$feature,cap=F),"s with larger variance)",sep=""),
                         to.plot=function() {
                           quant.heatmap(dolog2(quant.data.norm),do.cor=T,nlim=nlim,ColSideColors=coloursbygroup[conds])
                         }
                         )
    html <- paste(html,p$html,sep=" ")
    p <- gen.plot2report(filename=paste(file.prefix,"_norm_sample_heatmap_g2",nlim,".png",sep=""),
                       width=800,
                       height=800,
                       caption=paste("Heatmap: ",pprint.metric("nlib")," (",nlim,"  ",pprint.feat(opt$feature,cap=F),"s with larger variance)",sep=""),
                       to.plot=function() {
                         quant.heatmap(dolog2(quant.data.norm),do.cor=F,nlim=nlim,ColSideColors=coloursbygroup[conds])
                       }
                       )
    html <- paste(html,p$html,sep=" ")
  }
  irap.output.html(html,paste(heatmap.sample.info,sep=""))
  
} else {
  cat("<p>Insufficient data to generate heatmap</p>")
  pdebug("Insufficient data to generate heatmap")
}
%>

<H3><%=irap.ctr("Figure")%>: Principal components Analysis (PCA) applied to the expression values of all <%=pprint.feat(opt$feature,cap=F)%>s and all libraries</H3>  
<%
if (ncol(quant.data)>=2) {
  require("lattice")
  p <- gen.plot2report(filename=paste(file.prefix,"_sample_pca_g.png",sep=""),
                       width=600,
                       height=600,
                       caption="",
                       to.plot=function() {                           
                         pca <- prcomp(t(dolog2(quant.data)),center=T)
                         pca.sum <- summary(pca)
                         p <- xyplot(PC1 ~ PC2,data=as.data.frame(pca$x),
                                     cex=2,aspect="iso",
                                     col=coloursbysample[rownames(pca$x)],
                                     pch=16,
                                     xlab=paste("PC2 (",round(pca.sum$importance["Proportion of Variance","PC2"],2)*100,"% of variance)",sep=""),
                                     ylab=paste("PC1 (",round(pca.sum$importance["Proportion of Variance","PC1"],2)*100,"% of variance)",sep=""),
                                     main = suppressWarnings(draw.key(key = list(                             
                                                       rect = list(col = coloursbygroup),
                                                       text = list(names(coloursbygroup)),
                                                       columns=3
                                                       ))))
                         print(p)
                       }
                       )
  html <- p$html
  if ( opt$metric=="raw") {  
    p <- gen.plot2report(filename=paste(file.prefix,"_norm_sample_pca_g.png",sep=""),
                         width=600,
                         height=600,
                         caption=paste("PCA: ",pprint.metric("nlib")," (",nlim," ",pprint.feat(opt$feature,cap=F),"s with larger variance)",sep=""),

                         to.plot=function() {                           
                           pca <- prcomp(t(dolog2(quant.data.norm)),center=T)
                           pca.sum <- summary(pca)
                           p <- xyplot(PC1 ~ PC2,data=as.data.frame(pca$x),
                                       cex=2,aspect="iso",
                                       col=coloursbysample[rownames(pca$x)],
                                       pch=16,
                                       xlab=paste("PC2 (",round(pca.sum$importance["Proportion of Variance","PC2"],2)*100,"% of variance)",sep=""),
                                       ylab=paste("PC1 (",round(pca.sum$importance["Proportion of Variance","PC1"],2)*100,"% of variance)",sep=""),
                                       main = suppressWarnings(draw.key(key = list(                             
                                                         rect = list(col = coloursbygroup),
                                                         text = list(names(coloursbygroup)),
                                                         columns=3
                                                         ))))
                           print(p)
                         }
                         )
    html <- paste(html,p$html,sep=" ")    
  }
  irap.output.html(html,paste(" Principal Components Analysis (PCA) was applied to the matrix with the expression values by ",pprint.feat(opt$feature,cap=F)," and by sample.",sep=""))    
} else {
  cat("<p>Insufficient data to generate PCA</p>")
  pdebug("Insufficient data to generate PCA")
}
%>
    

</DIV>

<% brew::brew(get.path2template("page.footer")) %>

